### ISSUE TITLE
memory leaks

### ISSUE THREAD

**Issue by dralley** [RustPython/RustPython#2380]
I was running the microbenchmarks and kept getting SIGTERM, apparently it was exhausting my 16gb of memory + 8gb of swap.

I reran the tests to try to figure out which microbenchmarks were triggering the leaks in RustPython:

* exception_nested.py

![Screenshot from 2020-12-24 14-35-46](https://user-images.githubusercontent.com/2751996/103104777-47df7e80-45f7-11eb-8b55-90c0bb0aa192.png)

* complex_class.py

![Screenshot from 2020-12-24 14-37-51](https://user-images.githubusercontent.com/2751996/103104787-53cb4080-45f7-11eb-93bc-5637663aad06.png)

* define_function.py

![Screenshot from 2020-12-24 14-41-47](https://user-images.githubusercontent.com/2751996/103104793-5b8ae500-45f7-11eb-8a8c-665f5e8727b8.png)

* call_simple.py

![Screenshot from 2020-12-24 14-42-09](https://user-images.githubusercontent.com/2751996/103104804-6f364b80-45f7-11eb-8102-758f44c01925.png)

* call_kwargs.py

![Screenshot from 2020-12-24 14-44-16](https://user-images.githubusercontent.com/2751996/103104810-7fe6c180-45f7-11eb-937e-45ae79c46e23.png)

* construct_object.py

![Screenshot from 2020-12-24 14-46-25](https://user-images.githubusercontent.com/2751996/103104815-8aa15680-45f7-11eb-9d28-62e814e6f796.png)

This is the point where it died again, so maybe there are more.  Or maybe they all have the same root causes :shrug:

**Comment by carbotaniuman** [comment]
This is intentional, we don't have a GC implemented in RustPython yet, so any refernce cycles automatically leak memory.

**Comment by dralley** [comment]
Does simply calling or defining a function create a reference cycle, though?

**Comment by coolreader18** [comment]
Yes, I think, since the globals dict gets stored in the function object, and then the function object gets stored in the globals. I think we could get around that by clearing module globals on interpreter shutdown like CPython does, but that would require a defined "interpreter shutdown" phase.
